Initial pipelines docs (#18595)
cloudflare/cloudflare-docs: src/content/docs/pipelines/tutorials/send-data-from-client/index.mdx

---
updated: 2025-04-06
difficulty: Intermediate
content_type: 📝 Tutorial
pcx_content_type: tutorial
title: Create a data lake of clickstream data
products:
  - R2
  - Workers
tags:
  - MotherDuck
languages:
  - SQL
---

import { Render, PackageManagers, Details, WranglerConfig } from "~/components";

In this tutorial, you will learn how to build a data lake of website interaction events (clickstream data), using Pipelines.

Data lakes are a way to store large volumes of raw data in an object storage service such as [R2](/r2). You can run queries over a data lake to analyze the raw events.

> For this tutorial, you will build a landing page for an e-commerce website. Users can click on the website, to view products or add them to the cart. When a user clicks on buttons on the page, events will be sent to a pipeline. These events are sent from the client-side (directly from the user's browser). Your pipeline will automatically batch the ingested data, build output files, and deliver them to an [R2 bucket](/r2) to build your data lake.

Reviewer: ```suggestion
For this tutorial, you will build a landing page for an e-commerce website. Users can click on the website to view products or add them to the cart. When a user clicks on buttons on the page, events will be sent to a pipeline. These events are sent from the client-side (directly from the user's browser). Your pipeline will automatically batch the ingested data, build output files, and deliver them to an [R2 bucket](/r2) to build your data lake.
```

    				productElement.classList.add('rounded-lg', 'border', 'bg-card', 'text-card-foreground', 'shadow-sm');
    				productElement.innerHTML = `
            <div class="flex flex-col space-y-1.5 p-6">
              <h2 class="text-2xl font-semibold leading-none tracking-tight">${product.name}</h2>
            </div>
            <div class="p-6 pt-0">
              <p>${product.desc}</p>
              <p class="font-bold mt-2">$${product.cost}</p>
            </div>
            <div class="flex items-center p-6 pt-0 flex justify-between">
              <button class="border px-4 py-2 rounded-md" onclick="handleClick('product_view', ${product.id})" name="">View Details</button>
              <button class="border px-4 py-2 rounded-md" onclick="handleClick('add_to_cart', ${product.id})">Add to Cart</button>
            </div>
          `;
    				productContainer.appendChild(productElement);
    			});
    		}
    		renderProducts();

    		// function to handle click events
    		async function handleClick(action, id) {
    			console.log(`Clicked ${action} for product with id ${id}`);
    		}
    	</script>
    </body>

</html>
```
</details>

> The above code does the following:

Reviewer: ```suggestion
The above code:
```

- Uses Tailwind CSS to style the page.
- Renders a list of products.
- Adds a button to view the details of a product.
- Adds a button to add a product to the cart.
- Contains a `handleClick` function to handle the click events. This function logs the action and the product ID. In the next steps, you will create a pipeline and add the logic to send the click events to this pipeline.

## 3. Create an R2 Bucket

> We'll create a new R2 bucket to use as the sink for our pipeline. Create a new r2 bucket `clickstream-bucket` using the [Wrangler CLI](/workers/wrangler/). Open a terminal window, and run the following command:

Reviewer: ```suggestion
You will create a new R2 bucket to use as the sink for our pipeline. Create a new r2 bucket `clickstream-bucket` using the [Wrangler CLI](/workers/wrangler/). Open a terminal window and run the following command:
```

```sh
npx wrangler r2 bucket create clickstream-bucket
```

## 4. Create a pipeline
You need to create a new pipeline and connect it to your R2 bucket.

> Create a new pipeline `clickstream-pipeline-client` using the [Wrangler CLI](/workers/wrangler/). Open a terminal window, and run the following command:

Reviewer: ```suggestion
Create a new pipeline `clickstream-pipeline-client` using the [Wrangler CLI](/workers/wrangler/). Open a terminal window and run the following command:
```

```js title="public/index.html" ins={3-40}
async function handleClick(action, id) {
	console.log(`Clicked ${action} for product with id ${id}`);

	const userAgent = window.navigator.userAgent;
	const timestamp = new Date().toISOString();
	const { browser, os, device } = extractDeviceInfo(userAgent);

	const data = {
		timestamp,
		session_id: "1234567890abcdef", // For production use a unique session ID
		user_id: "user" + Math.floor(Math.random() * 1000), // For production use a unique user ID
		event_data: {
			event_id: Math.floor(Math.random() * 1000),
			event_type: action,
			page_url: window.location.href,
			timestamp,
			product_id: id,
		},
		device_info: {
			browser,
			os,
			device,
			userAgent,
		},
		referrer: document.referrer,
	};
	try {
		const response = await fetch("<PIPELINE_URL>", {
			method: "POST",
			headers: {
				"Content-Type": "application/json",
			},
			body: JSON.stringify([data]),
		});
		if (!response.ok) {
			throw new Error("Failed to send data to pipeline");
		}
	} catch (error) {
		console.error("Error sending data to pipeline:", error);
	}
}
```

> The `handleClick` function does the following:

Reviewer: ```suggestion
The `handleClick` function:
```

- Gets the device information using the `extractDeviceInfo` function.
- Makes a `POST` request to the pipeline with the data.
- Logs any errors that occur.

If you start the development server and open the application in the browser, you can see the `handleClick` function gets executed when you click on the `View Details` or `Add to Cart` button.

```sh
npm run dev
```

However, no data gets sent to the pipeline. Inspect the browser console to view the error message. The error message you see is for [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS). In the next step, you will update the CORS settings to allow the client-side JavaScript to send data to the pipeline.

## 7. Update CORS settings

By default, the HTTP ingestion endpoint for your pipeline does not allow cross-origin requests. You need to update the CORS settings to allow the client-side JavaScript to send data to the pipeline. To update the CORS settings, execute the following command:

```sh
npx wrangler pipelines update clickstream-pipeline-client --cors-origins http://localhost:8787
```

> Now when you run the development server locally, and open the website in a browser, clickstream data will be successfully sent to the pipeline. You can learn more about the CORS settings in the [Specifying CORS settings](/pipelines/build-with-pipelines/http/#specifying-cors-settings) documentation.

Reviewer: ```suggestion
Now when you run the development server locally and open the website in a browser, clickstream data will be successfully sent to the pipeline. You can learn more about the CORS settings in the [Specifying CORS settings](/pipelines/build-with-pipelines/http/#specifying-cors-settings) documentation.
```

## 8. Deploy the application

To deploy the application, run the following command:

```sh
npm run deploy
```

This will deploy the application to the Cloudflare Workers platform.

```output
🌀 Building list of assets...
🌀 Starting asset upload...
🌀 Found 1 new or modified static asset to upload. Proceeding with upload...
+ /index.html
Uploaded 1 of 1 assets
✨ Success! Uploaded 1 file (2.37 sec)

Total Upload: 25.73 KiB / gzip: 6.17 KiB
Worker Startup Time: 15 ms
Uploaded e-commerce-pipelines-client-side (11.79 sec)
Deployed e-commerce-pipelines-client-side triggers (7.60 sec)
  https://<URL>.workers.dev
Current Version ID: <VERSION_ID>
```

> We now need to update the pipeline's CORS settings again. This time, we'll include the URL of our newly deployed application. Run the command below, and replace `<URL>` with the URL of the application.

Reviewer: ```suggestion
Now you need to update the pipeline's CORS settings again. This time, you will include the URL of our newly deployed application. Run the command below and replace `<URL>` with the URL of the application.
```

```sh
npx wrangler pipelines update clickstream-pipeline-client --cors-origins http://localhost:8787 https://<URL>.workers.dev
```

Now, you can access the application at the deployed URL. When you click on the `View Details` or `Add to Cart` button, the clickstream data will be sent to your pipeline.

## 9. View the data in R2

You can view the data in the R2 bucket. If you are not signed in to the Cloudflare dashboard, sign in and navigate to the [R2 overview](https://dash.cloudflare.com/?to=/:account/r2/overview) page.

> Open the bucket you configured for your pipeline in Step 3. You can see files, representing the clickstream data. These files are newline delimited JSON files. Each row in a file represents one click event. Download one of the files, and open it in your preferred text editor to see the output:

Reviewer: ```suggestion
Open the bucket you configured for your pipeline in Step 3. You can see files representing the clickstream data. These files are newline delimited JSON files, and each row in a file represents one click event. Download one of the files and open it in your preferred text editor to see the output:
```

{"timestamp":"2025-04-06T16:24:31.879Z","session_id":"1234567890abcdef","user_id":"user750","event_data":{"event_id":756,"event_type":"product_view","page_url":"https://<URL>.workers.dev/","timestamp":"2025-04-06T16:24:31.879Z","product_id":4},"device_info":{"browser":"Chrome","os":"Linux","device":"Mobile","userAgent":"Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/134.0.0.0 Mobile Safari/537.36"},"referrer":""}
{"timestamp":"2025-04-06T16:24:33.978Z","session_id":"1234567890abcdef","user_id":"user333","event_data":{"event_id":467,"event_type":"product_view","page_url":"https://<URL>.workers.dev/","timestamp":"2025-04-06T16:24:33.978Z","product_id":6},"device_info":{"browser":"Chrome","os":"Linux","device":"Mobile","userAgent":"Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/134.0.0.0 Mobile Safari/537.36"},"referrer":""}
```

## 10. Optional: Connect a query engine to your R2 bucket and query the data

> Once you have collected the raw events in R2, you might want to query the events, to answer questions such as "how many events occurred?". You can connect a query engine, such as MotherDuck, to your R2 bucket.

Reviewer: ```suggestion
Once you have collected the raw events in R2, you might want to query the events to answer questions such as, "How many events occurred?". You can connect a query engine, such as MotherDuck, to your R2 bucket.
```

1. Log in to the MotherDuck dashboard and select your profile.
2. Navigate to the **Secrets** page.
3. Select the **Add Secret** button and enter the following information:

   - **Secret Name**: `Clickstream pipeline`
   - **Secret Type**: `Cloudflare R2`
   - **Access Key ID**: `ACCESS_KEY_ID` (replace with the Access Key ID)
   - **Secret Access Key**: `SECRET_ACCESS_KEY` (replace with the Secret Access Key)

4. Select the **Add Secret** button to save the secret.

### Query the data
In this step, you will query the data stored in the R2 bucket using MotherDuck.

1. Navigate back to the MotherDuck dashboard and select the **+** icon to add a new Notebook.
2. Select the **Add Cell** button to add a new cell to the notebook.

3. In the cell, enter the following query and select the **Run** button to execute the query:

```sql
SELECT count(*) FROM read_json_auto('r2://clickstream-bucket/**/*');
```

The query will return a count of all the events received.

## Conclusion

You have successfully created a Pipeline and used it to send clickstream data from the client. Through this tutorial, you've gained hands-on experience in:

> 1. Creating a Workers project, using static assets

Reviewer: ```suggestion
- Creating a Workers project using static assets.
```

> 2. Generating and capturing clickstream data

Reviewer: ```suggestion
- Generating and capturing clickstream data.
```

> 3. Setting up a pipeline to ingest data into R2

Reviewer: ```suggestion
- Setting up a pipeline to ingest data into R2.
```

> 4. Deploying the application to Workers

Reviewer: ```suggestion
- Deploying the application to Workers.
```

> 5. Using MotherDuck to query the data

Reviewer: ```suggestion
- Using MotherDuck to query the data.
```
